crates/oxc_linter/src/rules/import/no_barrel_file.rs

```diff
@@ -0,0 +1,100 @@
+use oxc_ast::{ast::Statement, AstKind};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error(
+    "eslint-plugin-import(no-barrel-file): \
+            Avoid barrel files, they slow down performance, \
+            and cause large module graphs with modules that go unused."
+)]
+#[diagnostic(severity(warning), help("For more information visit this link: <https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/>"))]
+struct NoBarrelFileDiagnostic(#[label] pub Span);
+
+// TODO: consider moving this to lint options. Biome doesn't proide this option,
+// The original eslint plugin does and it is 3 by default. I've set it to 0 for now to behave
+// similar to the biome.
+/// Minimum amount of exports to consider module as barrelfile
+const AMOUNT_OF_EXPORTS_TO_CONSIDER_MODULE_AS_BARREL: u8 = 0;
+
+/// <https://github.com/thepassle/eslint-plugin-barrel-files/blob/main/docs/rules/avoid-barrel-files.md>
+#[derive(Debug, Default, Clone)]
+pub struct NoBarrelFile;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow the use of barrel files.
+    ///
+    /// ### Example
+    ///
+    /// Invalid:
+    /// ```javascript
+    /// export { foo } from 'foo';
+    /// export { bar } from 'bar';
+    /// export { baz } from 'baz';
+    /// export { qux } from 'qux';
+    /// ```
+    /// Valid:
+    /// ```javascript
+    /// export type { foo } from './foo.js';
+    /// ```
+    NoBarrelFile,
+    nursery
+);
+
+impl Rule for NoBarrelFile {
+    fn run_once(&self, ctx: &LintContext<'_>) {
+        let semantic = ctx.semantic();
+        let mod_rec = semantic.module_record();
+        let root = semantic.nodes().root_node();
+
+        let AstKind::Program(program) = root.kind() else { unreachable!() };
+
+        let declarations = program.body.iter().fold(0, |acc, node| match node {
+            Statement::Declaration(_) => acc + 1,
+            _ => acc,
+        });
+        let exports = mod_rec.star_export_entries.len() + mod_rec.indirect_export_entries.len();
+
+        if exports > declarations
+            && exports > AMOUNT_OF_EXPORTS_TO_CONSIDER_MODULE_AS_BARREL as usize
+        {
+            ctx.diagnostic(NoBarrelFileDiagnostic(program.span));
+        }
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        // r#"export type * from "foo";"#,
```

Comment:
Why these don't pass? Are we lacking the correct syntax for the export type?

@Boshen